namespace/namespace.go

```diff
@@ -0,0 +1,268 @@
+)
+
+type endpoint struct {
+	action   string
+	url      *url.URL
+	priority int
+	flags    map[string]string
+}
+
+func (e *endpoint) Action() string {
+	return e.action
+}
+
+func (e *endpoint) BaseURL() *url.URL {
+	return e.url
+}
+
+func (e *endpoint) Priority() int {
+	return e.priority
+}
+
+func (e *endpoint) Flags() map[string]string {
+	return e.flags
+}
+
+func (e *endpoint) String() string {
+	return e.url.String()
+}
+
+var (
+	flagExp     = regexp.MustCompile("^([[:alnum:]][[:word:]]*)(?:=([[:graph:]]+))$")
+	priorityExp = regexp.MustCompile("^[[:digit:]]+$")
+)
+
+// createEndpoint parses the entry into a RemoteEndpoint.
+// The arguments are treated as <endpoint> [<priority>] [<flag>[=<value>]]...
+func createEndpoint(entry Entry) (distribution.RemoteEndpoint, error) {
+	if len(entry.args) == 0 {
+		return nil, errors.New("missing endpoint argument")
+	}
+	base, err := url.Parse(entry.args[0])
+	if err != nil {
+		return nil, err
+	}
+	// If scheme is empty, reparse with default scheme
+	if base.Scheme == "" {
+		base, err = url.Parse("https://" + entry.args[0])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	priority := 0
+	var prioritySet bool
+	flags := map[string]string{}
+	for _, arg := range entry.args[1:] {
+		if !prioritySet {
+			// Attempt match on
+			if priorityExp.MatchString(arg) {
+				i, err := strconv.Atoi(arg)
+				if err != nil {
+					return nil, err
+				}
+				priority = i
+				prioritySet = true
+				continue
+			}
+		}
+		matches := flagExp.FindStringSubmatch(arg)
+		if len(matches) == 2 {
+			flags[matches[1]] = ""
+		} else if len(matches) == 3 {
+			flags[matches[1]] = matches[2]
+		} else {
+			return nil, fmt.Errorf("invalid flag %q", arg)
+		}
+	}
+
+	return &endpoint{
+		action:   entry.action,
+		url:      base,
+		priority: priority,
+		flags:    flags,
+	}, nil
+}
+
+type resolvedNamespace struct {
+	scope       scope
+	ca          string
+	caScope     string
+	endpoints   []distribution.RemoteEndpoint
+	repoFactory RepositoryClientFactory
+}
+
+func (ns *resolvedNamespace) Scope() distribution.Scope {
+	return ns.scope
+}
+
+func (ns *resolvedNamespace) Repository(ctx context.Context, name string) (distribution.Repository, error) {
+	if !ns.scope.Contains(name) {
+		return nil, errors.New("repository out of namespace scope")
+	}
+
+	return ns.repoFactory(ctx, name, ns.endpoints)
+}
+
+type namespaceResolver struct {
+	entrySet    *Entries
+	discoverer  Discoverer
+	repoFactory RepositoryClientFactory
+}
+
+// RepositoryClientFactory is a factory function which will create a new
+// Repository object from the given endpoints.
+type RepositoryClientFactory func(ctx context.Context, namespace string, endpoints []distribution.RemoteEndpoint) (distribution.Repository, error)
+
+// NewNamespaceResolver returns a new Resolver for resolving namespaces from
+// the given namespace entry set.
+func NewNamespaceResolver(base *Entries, d Discoverer, rcf RepositoryClientFactory) Resolver {
```

Comment:
Instead of passing in a base set of Entries and a Discoverer, a Discoverer should be able to be composed of multiple discoverers.  The base entries should be able to be converted to a discoverer from a static set of entries and composed with any other methods.